Feature/backend

AndriiSerbeniuk opened the pull request:
Initiated rust crate for backend

AndriiSerbeniuk committed:
Updated gitignore to ignore target folders

AndriiSerbeniuk committed:
Init rust crate

virtustron committed:
Add package information to Compiler/Cargo.toml

virtustron reviewed (dismissed):
Its ok!

OleksandrStrashnenko reviewed (dismissed):
It is ok for now. 
Adding some rocket things would be great.

virtustron committed:
Add basic files to compiler crate

virtustron committed:
Trying to compile cpp

virtustron committed:
Add g++ compilation of cpp-file

virtustron committed:
Add todo

AndriiSerbeniuk committed:
Initialized basci rocket application

virtustron committed:
Make general structure of all modules

virtustron committed:
Add trait compiler

virtustron committed:
Fix Compiler trait implementation and usage

AndriiSerbeniuk commented:
Will be reopened when more features are implemented

AndriiSerbeniuk committed:
Added JSONs hadling

virtustron committed:
Simplify in and out data structures, fix errors and warnings

virtustron committed:
Change OutputData structure

virtustron committed:
Implement new out data structures in compile fun

virtustron committed:
Add handler::run_compilation usage to main fun

virtustron committed:
Add temp folder for source and compiled files

virtustron committed:
Implement compilation specific source file

virtustron committed:
Add filename generation by DateTime value

virtustron committed:
Write TODO list

AndriiSerbeniuk committed:
Added more of input data interpretation

AndriiSerbeniuk committed:
Tried to organize current codebase into modules

AndriiSerbeniuk committed:
Remade temp files handling

virtustron committed:
Add compiler_options to input_data and its usage

virtustron committed:
Moving to Path

AndriiSerbeniuk committed:
Simplified compiler output data struct

virtustron committed:
Use PathBuf type instead of Path

virtustron committed:
Add path to src and bin files to in and out data

virtustron committed:
Delete unnecessary bin file

AndriiSerbeniuk committed:
Merge branch 'feature/compiler' of https://github.com/ITA-Dnipro/CompilerWebApp into feature/backend

AndriiSerbeniuk committed:
Separated out the flework module

AndriiSerbeniuk committed:
Added compiler crate to the backend
Also:
- Changed filework functions to work with Paths;
- Renamed OutputCode to OutputData, InputCode to InputData;
- Implemented constructors for OutputData.

virtustron committed:
Add constructor to struct InputData

virtustron committed:
Add doc comment to run_compilation

virtustron committed:
Change stdout and stderr type to String

AndriiSerbeniuk committed:
Merge branch 'feature/compiler' of https://github.com/ITA-Dnipro/CompilerWebApp into feature/backend

AndriiSerbeniuk committed:
Adjusted backend code for new compiler code

valentynfaychuk commented on webapp/src/main.rs:
```diff
@@ -0,0 +1,21 @@
+    // Dir for compilation files
+    let mut temp_dir = current_dir().unwrap().to_str().unwrap().to_owned();
+    temp_dir.push_str("/tempdata");
+    set_var("COMPILATION_TEMP_DIR", temp_dir);
```
why not requesting `COMPILATION_TEMP_DIR` from outside?
```
match env::var(COMPILATION_TEMP_DIR) {
    Ok(dir) => {}, // possible validation if path exists, etc.
    Err(e) => {
        // initialize temp_dir
        println!("can't read COMPILATION_TEMP_DIR, using default path \"{}\"", temp_dir),
        set_var(COMPILATION_TEMP_DIR, temp_dir);
    }
}
```

valentynfaychuk commented on webapp/src/http_handlers/submit/mod.rs:
```diff
@@ -0,0 +1,34 @@
+    // TODO: validate input data: compilation options, language etc.
+
+    // Compilation goes here
+    let source_file = write_source_to_file(&compilation_json.code,
```
Is it possible to create a separate function `try_to_compile(code, ops) -> Result<..>`?

valentynfaychuk commented on webapp/src/filework/mod.rs:
```diff
@@ -0,0 +1,35 @@
+pub fn delete_file(filename: &Path)
+{
+    remove_file(filename)
+        .expect("Panic on file deletion");
```
Should we crash when it is impossible to remove a file (probably the file is locked or was already removed)?

valentynfaychuk commented on webapp/src/filework/mod.rs:
```diff
@@ -0,0 +1,35 @@
+        .expect("Panic on copying of source code into the temp file");
+    code_file.seek(SeekFrom::Start(0)).expect("Panic on temp file seek");
+    
+    drop(code_file);
```
Code file should be automatically dropped when going out of its scope, is there any need in doing this explicitly?

AndriiSerbeniuk replied:
I did this because I reckoned the compiler would need the file closed, but I didn't ask if it actually is so, my bad. If the compiler is okay with the file being open elsewhere then dropping it explicitly is redundant.

AndriiSerbeniuk replied:
Oh, I see now. Body of this function was once a part of http_handlers::submit::post_submit() and this drop migrated from there. It is completely redundant here, it will be removed.

valentynfaychuk commented on webapp/src/filework/mod.rs:
```diff
@@ -0,0 +1,35 @@
+    input_file_name.push_str("/");
+    input_file_name.push_str(
+        &chrono::Utc::now()
+        .format("compilation_input-%Y-%m-%d-%H-%M-%S-%f").to_string()
```
Is it possible to reuse `generate_filename` somehow?

virtustron replied:
> Is it possible to reuse `generate_filename` somehow?

I'm sorry, could you explain the word "reuse"? Did you mean using this function twice in one lib and another lib in the same time?

valentynfaychuk commented on compiler/Cargo.toml:
```diff
@@ -1 +1,14 @@
+path = "src/lib.rs"
+
+[[bin]]
+name = "testbin"
```
Do we need this crate to be a separate program?

virtustron replied:
> Do we need this crate to be a separate program?

Of course, no - compiler is a lib in fact. When we run "cargo build", we get lib. And "cargo run" allows to run main function in "src/bin/compiler_main.rs" (I have read about this possibility, when I prepared my presentation about crates and modules and tried to explore it).
This option gives us two benefits:
1. Possibility of direct Compiler debugging - we have "main" function in library.
2. Use case (example) of the Compiler usage by backend "src/bin/compiler_main.rs".

So, we can remove it anytime.

Are there (or will be in the future) some critical  issues with this approach?

valentynfaychuk commented on compiler/src/compilers/cpp_compiler.rs:
```diff
@@ -0,0 +1,47 @@
+                    "-o", &full_bin_filename.into_os_string().into_string().unwrap(), 
+                    input_data.source_code_file_path.to_str().unwrap()])
+                .output()
+                .expect("failed to execute process");
```
Do we need to crash when we can't execute a compiler?

virtustron replied:
> Do we need to crash when we can't execute a compiler?

Of course, no. We should find out existence of the specific compiler. 
Am I correctly understand your question?

valentynfaychuk commented on compiler/src/compilers/cpp_compiler.rs:
```diff
@@ -0,0 +1,47 @@
+}
+
+impl Compiler for CppCompiler {
+    fn compile(&self, input_data: &InputData) -> OutputData {
```
Does this function always succeed, why not returning `Result<..>`?

virtustron replied:
There can be a lot of option that may bring this function to unsuccessful result. For instance "lack of the specific compiler" mentioned in previous comment. So, it will be more efficient to return "Result<..>" in this case - just added corresponding TODO record.

valentynfaychuk commented on compiler/src/compilers/compiler.rs:
```diff
@@ -0,0 +1,7 @@
+use super::super::data::output_data::OutputData;
+
+
+pub trait Compiler {
```
Is it possible to use `Template` pattern to implement compiler?

virtustron replied:
Actually, trait Compiler is the attempt to use Template in Handler module:

https://github.com/ITA-Dnipro/CompilerWebApp/blob/e8b7cf7480209a77980003fa75a2d0de2418625e/compiler/src/handler/mod.rs#L76-L86

In Compiler Explorer there are also some kind of the "Parent" compiler - BaseCompiler 

https://github.com/compiler-explorer/compiler-explorer/blob/84fa2fe598d94733a281fcf9410912daa1d3de82/lib/base-compiler.js#L49

Our compiler uses another approach, but the same idea. 
Will we get any critical issues with such Template?

valentynfaychuk commented on compiler/src/data/input_data/mod.rs:
```diff
@@ -0,0 +1,46 @@
+    pub compiler_options: String,
+}
+
+impl InputData {
```
Members are already public, is there a need in constructor?

virtustron replied:
> Members are already public, is there a need in constructor?

Constructor was added a little bit later after struct for the more effective initialization in backend. And now, I suppose we need to make private these pub fields.  Just added corresponding TODO record.

valentynfaychuk commented:
Can we add newlines at the end of files (you can configure in VS code to do it automatically)?
Can we add unit-tests?

virtustron reviewed (approved):
Ok!
